### PR TITLE
Vectorization analysis for input fusion

### PR DESCRIPTION
**Pull request by yiqian1**
This patch updates the rock-regularize pass to annotate which input of GenericOp needs to be vectorized
and uses this information to do vectorization analysis for input fusions.

Closes https://github.com/ROCm/rocMLIR-internal/issues/1398

### CODE REVIEW

File: mlir/lib/Dialect/Rock/Transforms/Regularize.cpp [hunk 1]
```diff
@@ -148,6 +147,38 @@ struct RegularizeGenericRewritePattern
...
 };
 
+void AnnotateGenericOp(Operation *op, MLIRContext *ctx) {
+  if (auto lgop = dyn_cast<linalg::GenericOp>(op)) {
+    int64_t majorTensorSize = 0;
+    int32_t majorTensorIdx = -1;
+    int32_t inputIdx = 0;
+    size_t argIdx = -1;
+    for (auto inp : lgop.getInputs()) {
+      while (auto transform = inp.getDefiningOp<TransformOp>())
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/Regularize.cpp]: I'd `ViewLikeOpInterface` here for future-proofing but that's a nit.

File: mlir/lib/Dialect/Rock/Transforms/Regularize.cpp [hunk 2]
```diff
@@ -148,6 +147,38 @@ struct RegularizeGenericRewritePattern
...
+        inp = transform.getInput();
+
+      if (inp.isa<BlockArgument>()) {
+        auto arg = dyn_cast<BlockArgument>(inp);
+        auto shape = inp.getType().cast<ShapedType>().getShape();
+        int64_t argSize = 1;
+        for (size_t dim = 0; dim < shape.size(); dim++)
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/Regularize.cpp]: ... this is `getNumElements()`

File: mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp [hunk 3]
```diff
@@ -819,11 +819,17 @@ findPostFusionTransforms(Value buffer, Operation *currentUser) {
...
-        return failure();
-      }
-      candidate = genericOut;
+        if (!genericOp->hasAttr("majorTensorNumber")) {
+          LLVM_DEBUG(llvm::dbgs()
+                     << "[vectorization] can't analyze linalg.generic "
+                        "without majorTensorNumber\n");
+          return failure();
+        }
+        int32_t index =
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp]: `uint32_t` or `size_t`

File: mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp [hunk 4]
```diff
@@ -819,11 +819,17 @@ findPostFusionTransforms(Value buffer, Operation *currentUser) {
       }
       Value genericOut = genericOp.getOutputs().front();
       if (genericOut == buffer) {
-        LLVM_DEBUG(llvm::dbgs() << "[vectorization] Currently can't analyze "
-                                   "through input fusions\n");
-        return failure();
-      }
-      candidate = genericOut;
+        if (!genericOp->hasAttr("majorTensorNumber")) {
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp]: You  can 'getAttrOfType` and then test for `null` to save on the cast below

File: mlir/test/fusion/rock-gemm-align-tiling-multiple-args.mlir [hunk 5]
```diff
@@ -1,8 +1,10 @@
-// RUN: rocmlir-driver --rock-gridwise-gemm-to-blockwise --rock-blockwise-gemm-to-threadwise --rock-linalg-align --verify-passes %s | rocmlir-opt
+// RUN: rocmlir-driver --rock-regularize %s | FileCheck %s
+// RUN: rocmlir-driver --rock-regularize --rock-gridwise-gemm-to-blockwise --rock-blockwise-gemm-to-threadwise --rock-linalg-align --verify-passes %s | rocmlir-opt
```
Comment (krzysz00) [inline mlir/test/fusion/rock-gemm-align-tiling-multiple-args.mlir:2]: Add a `FileCheck` with a `--check-prefix=` maybe?

File: mlir/lib/Dialect/Rock/Transforms/Regularize.cpp [hunk 6]
```diff
@@ -148,6 +147,38 @@ struct RegularizeGenericRewritePattern
...
+            (argSize == majorTensorSize && argIdx > arg.getArgNumber())) {
+          majorTensorIdx = inputIdx;
+          majorTensorSize = argSize;
+          argIdx = arg.getArgNumber();
+        }
+      }
+      inputIdx++;
+    }
+    if (majorTensorIdx >= 0)
+      lgop->setAttr("majorTensorNumber",
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/Regularize.cpp]: Put a `rock.` on this

**Review by krzysz00 — COMMENTED** [review]
This does fix the common case but I'd think we want to handle the general case?

Also, before approval, can we get the PR message populated with a description of what's going on.

... Also, as a case, if there's one input argument, make that the major tensor even if it doesn't go to a function argument.

File: mlir/test/fusion/rock-gemm-align-tiling-multiple-args.mlir (hunk 5 above)
Comment (yiqian1) [reply mlir/test/fusion/rock-gemm-align-tiling-multiple-args.mlir:2]: I think this is OK. We have several test cases with FileCheck to validate the expected structure of IR. This test aims to verify that the IR doesn't violate SSA.

File: mlir/lib/Dialect/Rock/Transforms/Regularize.cpp [hunk 7]
```diff
@@ -148,6 +147,42 @@ struct RegularizeGenericRewritePattern
   }
 };
 
+void AnnotateGenericOp(Operation *op, MLIRContext *ctx) {
+  if (auto lgop = dyn_cast<linalg::GenericOp>(op)) {
+    int64_t majorTensorSize = 0;
+    int32_t majorTensorIdx = -1;
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/Regularize.cpp]: Nit: uint32_t on these for consistency

**Review by krzysz00 — APPROVED** [review]
Seems fine to me

**Comment by yiqian1** [comment]
This patch updates the rock-regularize pass to annotate which input of GenericOp needs to be vectorized
and uses this information to do vectorization analysis for input fusions.

**Pull request merged**